### DISCUSSION TITLE
deps before or after cmds?

### DISCUSSION THREAD

**Discussion by ssbarnea** [go-task/task#782] [Q&A]
I recently faced a dilemma related to the fact that the `cmds:` I had to run were supposed to run before `deps:`, which is not the normal behavior.

I was able to address this issue doing:

```yaml
  foo:
    cmds:
      - ...
      - task: bar
```

Still, I am almost sure that this approach prevents task from recognizing the dependency graph. I mention this because one feature I would like to build is ability to display display a graph, maybe with mermaid like:

```mermaid
flowchart LR
test --> build
build --> setup
package --> build
default --> lint
default --> package
default --> test
lint --> setup
```

Using deps to built such a graph would be easy, but if dependencies are hidden inside cmds, that task is no longer easy.

**Comment by andreynering** [comment]
Hi @ssbarnea,

May not be what you're looking for, but `defer` can be used to ensure something will be run at the end of the execution of the task:

https://taskfile.dev/usage/#doing-task-cleanup-with-defer

I know this won't solve the graph use case, though.

In theory you could have an option to consider calls like `task: foo` as part of the dependency graph just like `deps`. This would have the side-effect of including tasks that are not really dependencies, though. Doing that with `defer` would at least reduce the number of false-positives.

In summary, I don't have an exact solution for what you want. I also don't think I want to build a "`deps` but after `cmds`" feature just to address this graph use case.

**Reply by ssbarnea** [reply]
I agree the adding more complexity should be avoided. At least we have a thread now and we can decide how to address gathering these indirect dependencies when we are about to export the dependency graph.

I forgot to mention that I was hoping to include that info in the upcoming json export.
